src/audio/tdfb/tdfb.c

```diff
@@ -221,19 +341,25 @@ static int tdfb_setup(struct tdfb_comp_data *cd, int source_nch, int sink_nch)
 	if (!delay_size)
 		return 0;
 
-	/* Allocate all FIR channels data in a big chunk and clear it */
-	cd->fir_delay = rballoc(0, SOF_MEM_CAPS_RAM, delay_size);
-	if (!cd->fir_delay) {
-		comp_cl_err(&comp_tdfb, "tdfb_setup(), delay allocation failed for size %d",
-			    delay_size);
-		return -ENOMEM;
-	}
+	if (delay_size > cd->fir_delay_size) {
+		/* Free existing FIR channels data if it was allocated */
+		tdfb_free_delaylines(cd);
+
+		/* Allocate all FIR channels data in a big chunk and clear it */
+		cd->fir_delay = rballoc(0, SOF_MEM_CAPS_RAM, delay_size);
```

Comment:
do we need to check if this is already allocated? It appears this function is called from cmd logic so we would call it n times and this would leak memory.

Reply:
Can you double check? I may well miss something but tdfb_setup() is called from prepare() and optionally from copy(). The cmd() handler should not directly cause calls to this.... I hope.

Reply:
nevermind,i missed the free before, this is good